aws/resource_aws_eip.go

```diff
@@ -231,13 +231,27 @@ func resourceAwsEipUpdate(d *schema.ResourceData, meta interface{}) error {
 
 	// If we are updating an EIP that is not newly created, and we are attached to
 	// an instance or interface, detach first.
-	if (d.Get("instance").(string) != "" || d.Get("association_id").(string) != "") && !d.IsNewResource() {
+	disassociate := false
+	if !d.IsNewResource() {
+		if (d.Get("instance").(string) != "") && d.HasChange("instance") {
+			disassociate = true
+		} else if (d.Get("association_id").(string) != "") && (d.HasChange("network_interface") || d.HasChange("associate_with_private_ip")) {
```

Comment:
Nitpicks: As above - I'd omit the brackets and reverse ordering of `HasChange` and `Get`.